### ISSUE TITLE
Simple client/server example blocks forever

### ISSUE THREAD

**Issue by clarkmcc** [libp2p/rust-yamux#144]
It would be really nice to have an example of how to properly use this library (https://github.com/libp2p/rust-yamux/issues/61). I'm running into an issue where my code blocks forever, and I think I'm not quite understanding some invariant of the library, or maybe something in Rust itself. Here's what I'm trying to do
1. Client connects to server
2. Client sends hello message to server
3. Server responds with hello to client
4. Server initiates a stream with client
5. Server sends hello message to client
6. Client responds with hello message to server

The code seems to block when the client attempts to `open_stream` while the server is also blocked on `next_stream`. It seems like those two functions should unblock each other, but as I said, I must be missing something.
```
server: started listener
server: accepted connection
client: connected to server
client: opening stream on server
server: waiting for next stream
```

```rust
use anyhow::Result;
use async_compat::CompatExt;
use futures::{AsyncReadExt, AsyncWriteExt};
use tokio::{
    net::{TcpListener, TcpStream},
    task,
};
use yamux::{Config, Connection, Mode};

#[tokio::main]
async fn main() -> Result<()> {
    let handle1 = task::spawn(async { server().await.unwrap() });
    let handle2 = task::spawn(async { client().await.unwrap() });
    futures::future::join_all(vec![handle2, handle1]).await;
    Ok(())
}

async fn server() -> Result<()> {
    let listener = TcpListener::bind("127.0.0.1:8080").await?;
    println!("server: started listener");
    let cfg = Config::default();
    loop {
        let (socket, _) = listener.accept().await?;
        println!("server: accepted connection");
        let mut conn = Connection::new(socket.compat(), cfg.clone(), Mode::Server);
        let mut ctrl = conn.control();

        // Wait for the first stream
        println!("server: waiting for next stream");
        let mut stream1 = conn.next_stream().await?.unwrap();
        println!("server: got next stream");

        // Read the client hello
        let mut msg = String::default();
        stream1.read_to_string(&mut msg).await?;
        println!("got message from client: {}", msg);

        // Respond with the server's hello
        stream1.write_all("hello from server".as_bytes()).await?;

        // Create a server-initiated stream in response
        let mut stream2 = ctrl.open_stream().await?;
        stream2
            .write_all("hello from server again".as_bytes())
            .await?;

        // Close streams and cleanup
        stream2.close().await?;
        stream1.close().await?;
        ctrl.close().await?;
    }
}

async fn client() -> Result<()> {
    let socket = TcpStream::connect("127.0.0.1:8080")
        .await
        .expect("connecting to server");
    println!("client: connected to server");

    let cfg = Config::default();
    let mut conn = Connection::new(socket.compat(), cfg, Mode::Client);
    let mut ctrl = conn.control();

    // Open the first stream and send a hello
    println!("client: opening stream on server");
    let mut stream1 = ctrl.open_stream().await?;
    stream1.write_all("hello from client".as_bytes()).await?;
    println!("client: sent hello to server");

    // Get a response
    let mut msg = String::default();
    stream1.read_to_string(&mut msg).await?;
    println!("got message from server: {}", msg);

    // Wait for the server's second stream and read the response
    let mut stream2 = conn
        .next_stream()
        .await?
        .expect("accepting server-initiated stream");
    let mut msg = String::default();
    stream2.read_to_string(&mut msg).await?;
    println!("got another message from server: {}", msg);

    Ok(())
}
```

**Comment by thomaseizinger** [comment]
The documentation on `next_stream` has your answer :)

https://docs.rs/yamux/0.10.2/yamux/struct.Connection.html#method.next_stream

> This must be called repeatedly in order to make progress.

You have to call `next_stream` constantly for any stream to make progress.

The tests also show some usage of the API.

**Comment by clarkmcc** [comment]
@thomaseizinger haha, so I actually did read that and in my real-world project it is being used in a while loop but you're right, in this issue reproduction example, I was not following the proper behavior. I re-jiggered the example a bit to illustrate that even when calling `next_stream` in a loop, the client and server are still blocking.

Note that in this example, the expected behavior is that the client never opens a stream, only the server does and therefore I'm not calling `next_stream` on the server side. Is that a problem? What I'm trying to do in this example is:
1. Initiate a yamux connection
2. Once per second, the server will open a stream with the client
3. The server will write a message to the client
4. The client will read the message from the server
5. The client will write a message to the server
6. The server will read the message from the client
7. The server will close the stream
8. Jump to step 2 and repeat.

Based on what I can tell from the logs below, the program is blocked on these two lines, which seems weird. I guess I'm not sure why the program would block if one side is waiting for streams to be initiated and the other side is trying to initiate a stream. Anyways, I must be doing something wrong.

**Client**
```rust
while let Some(mut stream) = conn.next_stream().await? {
```

**Server**
```rust
let mut stream = ctrl.open_stream().await.expect("opening stream");
```

```
server: started listener
client: connected to server
client: waiting on server stream
server: accepted connection
server: opening stream
```

```rust
use anyhow::Result;
use async_compat::CompatExt;
use futures::{AsyncReadExt, AsyncWriteExt};
use std::time::Duration;
use tokio::{
    net::{TcpListener, TcpStream},
    task,
    time::sleep,
};
use yamux::{Config, Connection, Mode};

#[tokio::main]
async fn main() -> Result<()> {
    let handle1 = task::spawn(async { server().await.unwrap() });
    let handle2 = task::spawn(async { client().await.unwrap() });
    futures::future::join_all(vec![handle2, handle1]).await;
    Ok(())
}

async fn server() -> Result<()> {
    let listener = TcpListener::bind("127.0.0.1:8080").await?;
    println!("server: started listener");
    loop {
        let cfg = Config::default();
        let (socket, _) = listener.accept().await?;
        println!("server: accepted connection");

        // Wait for client streams, for each client stream, receive the
        // client message, then start a server-initiated stream.
        task::spawn(async move {
            let conn = Connection::new(socket.compat(), cfg.clone(), Mode::Server);
            let mut ctrl = conn.control();

            loop {
                // Open stream
                println!("server: opening stream");
                let mut stream = ctrl.open_stream().await.expect("opening stream");
                println!("server: opened stream");

                // Send server hello
                stream
                    .write_all("hello from server".as_bytes())
                    .await
                    .expect("writing to client");

                // Read client response
                let mut str = String::default();
                stream
                    .read_to_string(&mut str)
                    .await
                    .expect("reading client hello");
                println!("got message from client: {}", str);

                // Close stream
                stream.close().await.expect("closing stream");

                sleep(Duration::from_secs(1)).await;
            }
        });
    }
}

async fn client() -> Result<()> {
    let socket = TcpStream::connect("127.0.0.1:8080")
        .await
        .expect("connecting to server");
    println!("client: connected to server");

    let cfg = Config::default();
    let mut conn = Connection::new(socket.compat(), cfg, Mode::Client);

    println!("client: waiting on server stream");
    while let Some(mut stream) = conn.next_stream().await? {
        println!("client: got server stream");

        let mut str = String::default();
        stream.read_to_string(&mut str).await?;
        println!("client: got message from server: {}", str);

        stream.write_all("hello from client".as_bytes()).await?;
        println!("client: responded to server stream");
    }
    println!("client: shutdown");
    Ok(())
}
```

**Comment by thomaseizinger** [comment]
> Note that in this example, the expected behavior is that the client never opens a stream, only the server does and therefore I'm not calling `next_stream` on the server side. Is that a problem?

Yes. You need to call `next_stream for any stream to make progress.

A "server" or any peer really always has to handle incoming streams.

With yamux, both parties can open streams so both parties need to have a "server"-loop that deals with incoming streams.

**Comment by clarkmcc** [comment]
Based on what I'm hearing here, it's not possible for me to replicate the [Go example](https://github.com/hashicorp/yamux#usage), is that correct? In other words, it doesn't seem like I can open a single stream on one side, accept a single stream on the other side, then write a message on one side, and read the message from the other side.

This behavior makes it difficult to have one control stream (for sending command messages) and several worker streams (for tunneling raw TCP traffic for example) because streams have to be received in a loop, rather than a pre-defined set of streams created and accepted during the initial connection. Pardon my ignorance, but it seems like it would be a better API to allow callers to drive the server loop separately from accepting connections. At first glance, I think there might be some borrowing issues with that approach since `next_stream` takes `&mut self`. But something like this seems more ergonomic to me personally.

---
Making progress, but now it's blocked on client reads. 

```
server: started listener
server: accepted connection
client: connected to server
client: waiting on server stream
server: opening stream
server: opened stream
server: sent message to client
client: got server stream
```

```rust
use anyhow::Result;
use async_compat::CompatExt;
use futures::{AsyncReadExt, AsyncWriteExt};
use std::time::Duration;
use tokio::{
    net::{TcpListener, TcpStream},
    task,
    time::sleep,
};
use yamux::{Config, Connection, Mode};

#[tokio::main]
async fn main() -> Result<()> {
    let handle1 = task::spawn(async { server().await.unwrap() });
    let handle2 = task::spawn(async { client().await.unwrap() });
    futures::future::join_all(vec![handle2, handle1]).await;
    Ok(())
}

async fn server() -> Result<()> {
    let listener = TcpListener::bind("127.0.0.1:8080").await?;
    println!("server: started listener");
    loop {
        let cfg = Config::default();
        let (socket, _) = listener.accept().await?;
        println!("server: accepted connection");

        // Wait for client streams, for each client stream, receive the
        // client message, then start a server-initiated stream.
        task::spawn(async move {
            let mut conn = Connection::new(socket.compat(), cfg.clone(), Mode::Server);
            let mut ctrl = conn.control();

            task::spawn(async move {
                while let Some(_) = conn.next_stream().await.unwrap() {
                    println!("client started a stream");
                }
            });

            loop {
                // Open stream
                println!("server: opening stream");
                let mut stream = ctrl.open_stream().await.expect("opening stream");
                println!("server: opened stream");

                // Send server hello
                stream
                    .write_all("hello from server".as_bytes())
                    .await
                    .expect("writing to client");
                println!("server: sent message to client");

                // Read client response
                let mut str = String::default();
                stream
                    .read_to_string(&mut str)
                    .await
                    .expect("reading client hello");
                println!("server: got message from client: {}", str);

                // Close stream
                stream.close().await.expect("closing stream");

                sleep(Duration::from_secs(1)).await;
            }
        });
    }
}

async fn client() -> Result<()> {
    let socket = TcpStream::connect("127.0.0.1:8080")
        .await
        .expect("connecting to server");
    println!("client: connected to server");

    let cfg = Config::default();
    let mut conn = Connection::new(socket.compat(), cfg, Mode::Client);

    println!("client: waiting on server stream");
    while let Some(mut stream) = conn.next_stream().await? {
        println!("client: got server stream");

        let mut str = String::default();
        stream.read_to_string(&mut str).await?;
        println!("client: got message from server: {}", str);

        stream.write_all("hello from client".as_bytes()).await?;
        println!("client: responded to server stream");
    }
    println!("client: shutdown");
    Ok(())
}
```

**Comment by clarkmcc** [comment]
I spent some time testing out [quinn](https://github.com/quinn-rs/quinn) (a QUIC library) this morning and it seems to have the semantics that I'm looking for in my project. Thanks for going back and forth with me on this.

**Comment by thomaseizinger** [comment]
> Based on what I'm hearing here, it's not possible for me to replicate the [Go example](https://github.com/hashicorp/yamux#usage), is that correct?

It is possible to replicate, but contrary to Go, Rust doesn't ship with a runtime so we cannot just do things magically in the background :)

> In other words, it doesn't seem like I can open a single stream on one side, accept a single stream on the other side, then write a message on one side, and read the message from the other side.

You can do that, you just need to spawn a "server" on either end that drives the connection whilst you want to write to this stream!

I spent some time yesterday refactoring the tests, perhaps that makes things a bit easier to understand?

https://github.com/libp2p/rust-yamux/pull/145

> This behavior makes it difficult to have one control stream (for sending command messages) and several worker streams (for tunneling raw TCP traffic for example) because streams have to be received in a loop, rather than a pre-defined set of streams created and accepted during the initial connection.

This is possible and probably best modelled with an enum that represents the various states of the connection setup. I'll try and put a code sample together that shows how :)

> Pardon my ignorance, but it seems like it would be a better API to allow callers to drive the server loop separately from accepting connections.

Yes I agree that the API needs to improve! One issue is that yamux unfortunately doesn't have a backpressure mechanism in the spec that allows a peer to control, how many streams the remote can open. A peer just needs to send a frame with a `SYN` flag and bam, _a wild stream appears_.

As a result, it is literally outside of our control, when the remote opens a stream and it just happens as part of reading data from the socket.

You can make a channel with a background task that drives the connection and allows you make progress without consuming the streams at the same place. But bear in mind that if that channel is unbounded, you have a DoS risk and if it is bounded, your connection may stall again if you don't consume those incoming streams.

> At first glance, I think there might be some borrowing issues with that approach since `next_stream` takes `&mut self`.

Rust's mutability and ownership rules certainly make it more challenging to design something here because we can't just willy-nilly share data :)

`async` and `await` don't play well with that either which is why I am removing the `async` style driver and control API in #142.

**Comment by zerolfx** [comment]
I spent a few hours on the same problem and finally found this issue.

It would be helpful to include an example in the [Connection](https://docs.rs/yamux/latest/yamux/struct.Connection.html) document, clarifying that `poll_next_inbound` is necessary even if the connection only serves as a client.

**Comment by thomaseizinger** [comment]
> It would be helpful to include an example in the [Connection](https://docs.rs/yamux/latest/yamux/struct.Connection.html) document, clarifying that `poll_next_inbound` is necessary even if the connection only serves as a client.

There is no such thing as "only serves as a client in yamux". The yamux protocol only distinguishes between client and server in how they allocate stream IDs but there is no capabilities on what either end of the connection can do. Consequently, a library that implements the protocol MUST be designed to handle a new inbound stream at all times.

I am open to PRs that extend the documentation although it pretty clearly states that `poll_next_inbound` must be called to make progress: https://github.com/libp2p/rust-yamux/blob/618b141786ea29fa5b7617cd2d97544e18a67b00/yamux/src/connection.rs#L76-L78
